File: day-6/day-6-1.py
#!/usr/bin/env python


def checksum(orb_map):
    def helper(m, depth, k):
        if m.get(k):
            return depth + sum([helper(m, depth + 1, cs) for cs in m.get(k)])
        else:
            return depth

    return helper(orb_map, 0, 'COM')


def add_to_map(orb, orb_map):
    parent, child = orb[:-1].split(')')

    if orb_map.get(parent):
        orb_map[parent].append(child)
    else:
        orb_map[parent] = [child]


def parse(ipt):
    orb_map = {}

    for i in ipt:
        add_to_map(i, orb_map)

    return orb_map


def solve(ipt):
    return checksum(parse(ipt))


if __name__ == '__main__':
    with open('day-6/input.txt') as inputfile:
        print(solve(inputfile.readlines()))

File: day-6/day-6-1.hs
import qualified Data.Map.Lazy as Map
import Debug.Trace

main = do
    rawInput <- readFile "input.txt"
    let input = lines rawInput
     in print $ solve input

solve orbList =
     let orbMap = foldl addToMap Map.empty orbList
     in traverseOrbs orbMap

addToMap orbMap orb =
    let (keyBrkt,val) = splitAt 4 orb
        key = take 3 keyBrkt
     in case Map.lookup key orbMap of
          Just v -> Map.insert key (val : v) orbMap
          Nothing -> Map.insert key [val] orbMap

traverseOrbs orbMap =
    let traverseOrbs depth k =
            case Map.lookup k orbMap of
              Just v -> (+) depth $ sum $ map (traverseOrbs $ depth + 1) v
              Nothing -> depth
     in traverseOrbs 0 "COM"
